### PR TITLE
Port some code to Python3

### PR DESCRIPTION
**Pull request by ghostd**
With this PR i wish to share modifications to run `mach` with `./mach` instead of `python3 ./mach`.

I'm using a Debian machine. The following commands work for me:
* `./mach build --dev`
* `./mach test-tidy`
* `./mach fmt`

I still have errors with the WPT scripts, but maybe the root causes is a bad resolution of `requirements.txt`

I used `2to3` to migrate some code before applying some manual fixes. For PLY, I copied/paste the code from the github repository.

If someone can test...
For now i'm not sure how to continue the migration while i can't run WPT :-/

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by ghostd** [comment]
I updated and rebased the PR. My [last issue](https://github.com/web-platform-tests/wpt/issues/27025) needs a [WPT fix](https://github.com/web-platform-tests/wpt/pull/26976) (and is specific to Python 3.9)

I can run bootstrap, build, test-wpt, fmt and test-tidy.

**Comment by jdm** [comment]
```
Running the WPT lint on ./tests/wpt/web-platform-tests/...
{"path": "import-maps/data-driven/tools/format_json.py", "lineno": 15, "rule": "PARSE-FAILED", "message": "Unable to parse file"}
```
This is an interesting failure.

**Comment by jdm** [comment]
That presumably comes from https://github.com/web-platform-tests/wpt/blob/98049d97b75b4edbc71ee857b2b424bfec82524d/tools/lint/lint.py#L687-L696.

**Review by jdm — APPROVED** [review]
The code changes look fine to me.

**Comment by jdm** [comment]
We should update https://github.com/servo/servo/blob/master/etc/taskcluster/decision_task.py to always use python3.

**Comment by ghostd** [comment]
> ```
> Running the WPT lint on ./tests/wpt/web-platform-tests/...
> {"path": "import-maps/data-driven/tools/format_json.py", "lineno": 15, "rule": "PARSE-FAILED", "message": "Unable to parse file"}
> ```
> 
> This is an interesting failure.

I proposed a [PR](https://github.com/web-platform-tests/wpt/pull/27037) to fix it

**Comment by ghostd** [comment]
> We should update https://github.com/servo/servo/blob/master/etc/taskcluster/decision_task.py to always use python3.

I added `python3` for each `mach` command and tried to add a `with_python3` method for the Windows build (but i have no clue if it works). This commit really needs a careful review :-)

File: etc/taskcluster/decisionlib.py [hunk 1]
```diff
@@ -637,6 +637,24 @@ def with_python2(self):
...
+        """
+        return self \
+            .with_directory_mount(
+            "https://www.python.org/ftp/python/3.7.3/python-3.7.3-embed-amd64.zip",
+            sha256="6de14c9223226cf0cd8c965ecb08c51d62c770171a256991b4fddc25188cfa8e",
+            path="python3",
+        ) \
+        .with_early_script("""
+            python -m ensurepip
+            pip install virtualenv==20.2.1
```
Comment (jdm) [inline etc/taskcluster/decisionlib.py:655]: I think these may need to be `python3` and `pip3` respectively.

File: etc/taskcluster/decisionlib.py [hunk 2]
```diff
@@ -637,6 +637,24 @@ def with_python2(self):
...
+        """)
```
Comment (jdm) [inline etc/taskcluster/decisionlib.py:656]: I believe this also requires `.with_path_from_homedir("python3", "python3\\Scripts")`

**Comment by tomoverlund** [comment]
How did you get ./mach build --dev without changes to mozjs?

**Comment by jdm** [comment]
Presumably it was still using python2 for the mozjs build.

**Comment by tomoverlund** [comment]
> Presumably it was still using python2 for the mozjs build.

Oh, that makes sense. I'm trying to build in Debian unstable with no python2 in sight.

File: etc/taskcluster/decisionlib.py (hunk 1 above)
Comment (ghostd) [reply etc/taskcluster/decisionlib.py]: Done

File: etc/taskcluster/decisionlib.py (hunk 2 above)
Comment (ghostd) [reply etc/taskcluster/decisionlib.py:660]: Done

**Comment by ghostd** [comment]
I rebased and squashed the commit. The WPT PR has been merged \o/

**Comment by jdm** [comment]
I'm working on incorporating these changes into #28092.

**Comment by jdm** [comment]
I'm going to attempt to wrestle CI into shape in #28141.

**Comment by jdm** [comment]
I'm moving ahead with merging this code as part of #28092, because my attempts to get Windows CI working in #28141 didn't work out. I'm going to go ahead and close this PR—thank you so much for doing this work! It was a big help.

**Pull request closed without merging**